src/main/java/frc/robot/subsystem/DriveSubsystem.java

```diff
@@ -0,0 +1,272 @@
+                if (turnRequest != 0) {
+                    turnState = TurnState.TURNING;
+                }
+                break;
+            case TURNING:
+                if (turnRequest == 0 && Math.abs(gyro.getRate()) < TURN_RATE_CUTOFF) {
+                    maintainHeading = currentHeading;
+                    turnState = TurnState.HOLD_HEADING;
+                }
+                break;
+        }
+        switch (turnState) {
+            case HOLD_HEADING:
+                if (shouldHeadingPID.getBoolean(true)) {
+                    return headingController.calculate(
+                            currentHeading.getRadians(), maintainHeading.getRadians());
+                }
+            case TURNING:
+            default:
+                return turnRequest;
+        }
+    }
+
+    public void driveCartesian(double fwdRequest, double leftRequest, double turnRequest) {
+        this.driveCartesian(fwdRequest, leftRequest, turnRequest, new Rotation2d());
+    }
+
+    public void driveCartesian(
+            double fwdRequest,
+            double leftRequest,
+            double turnRequest,
+            Rotation2d fieldOrientedDriveAngle) {
+        turnRequest = calculateTurn(turnRequest, getCurrHeading());
+
+        double maxLinearSpeed = MoPrefs.maxDriveSpeed.get();
+        double maxAngularSpeed = MoPrefs.maxTurnSpeed.get();
+
+        ChassisSpeeds speeds =
+                ChassisSpeeds.fromFieldRelativeSpeeds(
+                        -fwdRequest * maxLinearSpeed,
+                        leftRequest * maxLinearSpeed,
+                        turnRequest * maxAngularSpeed,
+                        fieldOrientedDriveAngle);
+
+        driveSwerveStates(kinematics.toSwerveModuleStates(speeds));
+    }
+
+    public void stop() {
+        frontLeft.driveMotor.stopMotor();
+        frontRight.driveMotor.stopMotor();
+        rearLeft.driveMotor.stopMotor();
+        rearRight.driveMotor.stopMotor();
+    }
+
+    public void driveSwerveStates(SwerveModuleState[] states) {
+        SwerveDriveKinematics.desaturateWheelSpeeds(states, MoPrefs.maxDriveSpeed.get());
+
+        frontLeft.drive(states[0]);
+        frontRight.drive(states[1]);
+        rearLeft.drive(states[2]);
+        rearRight.drive(states[3]);
+    }
+
+    public boolean isMoving() {
+        return (frontLeft.driveMotor.getSelectedSensorVelocity() / MoPrefs.flDriveMtrScale.get())
+                        > MOVE_RATE_CUTOFF
+                || (frontRight.driveMotor.getSelectedSensorVelocity()
+                                / MoPrefs.frDriveMtrScale.get())
+                        > MOVE_RATE_CUTOFF
+                || (rearLeft.driveMotor.getSelectedSensorVelocity() / MoPrefs.rlDriveMtrScale.get())
+                        > MOVE_RATE_CUTOFF
+                || (rearRight.driveMotor.getSelectedSensorVelocity()
+                                / MoPrefs.rrDriveMtrScale.get())
+                        > MOVE_RATE_CUTOFF;
+    }
```

Comment:
Do we use this function?

Reply:
Yes, we do not accept any april tag positioning if the robot is moving past the cutoff rate